tools/depcheck/pkg/cmd/trace/trace.go

```diff
@@ -0,0 +1,301 @@
+
+	r, w := io.Pipe()
+	golist.Stdout = w
+	golist.Stderr = os.Stderr
+
+	pkgs := &PackageList{}
+	go func(list *PackageList) {
+		decoder := json.NewDecoder(r)
+		for {
+			var pkg Package
+			err := decoder.Decode(&pkg)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+			}
+
+			list.Add(pkg)
+		}
+	}(pkgs)
+
+	if err := golist.Run(); err != nil {
+		return err
+	}
+
+	g, err := BuildGraph(pkgs, o.excludes)
+	if err != nil {
+		return err
+	}
+
+	if len(o.filters) > 0 {
+		g, err = FilterPackages(g, o.filters)
+		if err != nil {
+			return err
+		}
+	}
+
+	// prune orphans after filtering in order to maximize chance of deleting disconnected
+	// trees from the graph that have been collapsed into their root node by a given filter.
+	g.PruneOrphans()
+
+	if len(o.OutputFormat) > 0 {
+		return o.outputGraph(g)
+	}
+
+	if len(o.Analyze) > 0 {
+		return o.analyzeGraph(g)
+	}
+
+	return nil
+}
+
+func (o *TraceImportsOpts) outputGraph(g graph.Directed) error {
+	if o.OutputFormat != "dot" {
+		return fmt.Errorf("invalid output format provided: %s", o.OutputFormat)
+	}
+
+	data, err := dot.Marshal(g, fmt.Sprintf("%q", strings.Join(o.Roots, ", ")), "", " ", false)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(o.Out, "%v\n", string(data))
+	return nil
+}
+
+// analyzeGraph receives a MutableDirectedGraph and outputs
+// - "Yours": a list of every node in the set unique to the dependency tree of a given target node.
+// - "Mine": a list of every node in the set unique to the dependency tree of the root nodes (set non-overlapping with given target node)
+// - "Ours": a list of every node in the overlapping set between the dependency tree of the root nodes and a given target node
+func (o *TraceImportsOpts) analyzeGraph(g *depgraph.MutableDirectedGraph) error {
+	nodeToAnalyze, exists := g.NodeByName(o.Analyze)
+	if !exists {
+		return fmt.Errorf("unable to find package with import path %q", o.Analyze)
+	}
+	you, ok := nodeToAnalyze.(*depgraph.Node)
+	if !ok {
+		return fmt.Errorf("expected node to analyze to be of type *depgraph.Node. Got: %v", you)
+	}
+
+	fmt.Printf("Analyzing a total of %v packages\n", len(g.Nodes()))
+	fmt.Println()
+
+	yours := calculateOrphans(g, []*depgraph.Node{you})
+	fmt.Printf("\"Yours\": %v packages exclusive to %q\n", len(yours), you.UniqueName)
+	for _, n := range yours {
+		fmt.Printf("    - %s\n", n.UniqueName)
+	}
+
+	return nil
+}
+
+// calculateOrphans receives a graph and a set of targetNodes.
+// The graph is cloned, and all targetNodes are removed from it.
+// Any nodes left orphaned are added to a "removed" set.
+// This operation is done recursively until we no longer cause
+// any nodes to become orphaned after pruning the removed set.
+// The resulting set of total nodes recursively removed is returned.
+func calculateOrphans(g *depgraph.MutableDirectedGraph, targetNodes []*depgraph.Node) []*depgraph.Node {
```

Comment:
I think this is the graph method I want.  Peer to the other graph methods I'd think.  Unit tests located there to check it.